File: backend/routers/sucesion.py
"""Router de sucesión — mapa de talento y planes de carrera. empresa_id: lecturas por X-Empresa-Id (None = todas); CREATE heredada del empleado en el service."""
from uuid import UUID

from fastapi import APIRouter, Depends, Request

from schemas.sucesion import (
    EmpleadoAnalisisResponse, EmpleadoMapaResponse,
    HitoBodyCreate, HitoResponse,
    PlanCarreraCreate, PlanCarreraResponse, ReadinessUpdate,
)
from services.sucesion_service import SucesionService
from utils.empresa import get_empresa_id
from utils.permisos import Accion, Seccion, require_permission

router = APIRouter()
SECCION = Seccion.SUCESION


def _svc() -> SucesionService:
    return SucesionService()


@router.get("/mapa", response_model=list[EmpleadoMapaResponse], dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def get_mapa_talento(request: Request, svc: SucesionService = Depends(_svc)) -> list[EmpleadoMapaResponse]:
    return svc.get_mapa_talento(get_empresa_id(request))


@router.get("/analisis", response_model=list[EmpleadoAnalisisResponse], dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def get_analisis_posicion(
    request: Request, area_id: UUID, posicion: str = "", svc: SucesionService = Depends(_svc),
) -> list[EmpleadoAnalisisResponse]:
    return svc.get_analisis_posicion(area_id, posicion, get_empresa_id(request))


@router.get("/planes", response_model=list[PlanCarreraResponse], dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def get_planes_carrera(request: Request, svc: SucesionService = Depends(_svc)) -> list[PlanCarreraResponse]:
    return svc.get_planes_carrera(get_empresa_id(request))


@router.post("/planes", response_model=PlanCarreraResponse, status_code=201, dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def create_plan_carrera(
    data: PlanCarreraCreate, svc: SucesionService = Depends(_svc),
) -> PlanCarreraResponse:
    return svc.create_plan_carrera(data)


@router.get("/planes/{plan_id}/hitos", response_model=list[HitoResponse], dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def get_hitos(
    plan_id: UUID, svc: SucesionService = Depends(_svc),
) -> list[HitoResponse]:
    return svc.get_hitos(plan_id)


@router.post("/planes/{plan_id}/hitos", response_model=HitoResponse, status_code=201, dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def create_hito(
    plan_id: UUID, data: HitoBodyCreate, svc: SucesionService = Depends(_svc),
) -> HitoResponse:
    return svc.create_hito(plan_id, data)


@router.put("/planes/{plan_id}/readiness", response_model=PlanCarreraResponse, dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def update_readiness(
    plan_id: UUID, data: ReadinessUpdate, svc: SucesionService = Depends(_svc),
) -> PlanCarreraResponse:
    return svc.update_readiness(plan_id, data.readiness)


@router.put("/hitos/{hito_id}/completar", response_model=dict, dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def completar_hito(
    hito_id: UUID, svc: SucesionService = Depends(_svc),
) -> dict:
    svc.completar_hito(hito_id)
    return {"ok": True}

File: frontend/services/sucesion.ts
import { apiFetch } from "@/services/api"
import type {
  EmpleadoAnalisis, EmpleadoMapa,
  Hito, HitoCreate,
  PlanCarrera, PlanCarreraCreate,
} from "@/types/sucesion"

export async function fetchMapaTalento(): Promise<EmpleadoMapa[]> {
  return apiFetch<EmpleadoMapa[]>("/api/sucesion/mapa")
}

export async function fetchPlanesCarrera(): Promise<PlanCarrera[]> {
  return apiFetch<PlanCarrera[]>("/api/sucesion/planes")
}

export async function createPlanCarrera(data: PlanCarreraCreate): Promise<PlanCarrera> {
  return apiFetch<PlanCarrera>("/api/sucesion/planes", {
    method: "POST",
    body: JSON.stringify(data),
  })
}

export async function fetchHitos(planId: string): Promise<Hito[]> {
  return apiFetch<Hito[]>(`/api/sucesion/planes/${planId}/hitos`)
}

export async function createHito(planId: string, data: HitoCreate): Promise<Hito> {
  return apiFetch<Hito>(`/api/sucesion/planes/${planId}/hitos`, {
    method: "POST",
    body: JSON.stringify(data),
  })
}

export async function updateReadiness(planId: string, readiness: number): Promise<PlanCarrera> {
  return apiFetch<PlanCarrera>(`/api/sucesion/planes/${planId}/readiness`, {
    method: "PUT",
    body: JSON.stringify({ readiness }),
  })
}

export async function completarHito(hitoId: string): Promise<void> {
  await apiFetch<{ ok: boolean }>(`/api/sucesion/hitos/${hitoId}/completar`, {
    method: "PUT",
  })
}

export async function fetchAnalisisPosicion(
  areaId: string,
): Promise<EmpleadoAnalisis[]> {
  const params = new URLSearchParams({ area_id: areaId })
  return apiFetch<EmpleadoAnalisis[]>(`/api/sucesion/analisis?${params}`)
}
